crates/router/src/core/payment_link.rs

```diff
@@ -287,16 +276,89 @@ fn validate_order_details(
     Ok(updated_order_details)
 }
 
-fn extract_payment_link_config(
-    pl_config: Option<serde_json::Value>,
-) -> Result<Option<admin_types::PaymentLinkConfig>, error_stack::Report<errors::ApiErrorResponse>> {
-    pl_config
-        .map(|config| {
-            serde_json::from_value::<admin_types::PaymentLinkConfig>(config)
-                .into_report()
-                .change_context(errors::ApiErrorResponse::InvalidDataValue {
-                    field_name: "payment_link_config",
-                })
+pub fn extract_payment_link_config(
+    pl_config: serde_json::Value,
+) -> Result<api_models::admin::PaymentLinkConfig, error_stack::Report<errors::ApiErrorResponse>> {
+    serde_json::from_value::<api_models::admin::PaymentLinkConfig>(pl_config.clone())
+        .into_report()
+        .change_context(errors::ApiErrorResponse::InvalidDataValue {
+            field_name: "payment_link_config",
         })
-        .transpose()
+}
+
+pub fn get_payment_link_config_based_on_priority(
+    payment_create_link_config: Option<api_models::payments::PaymentCreatePaymentLinkConfig>,
+    business_link_config: Option<serde_json::Value>,
+    merchant_name: String,
+    default_domain_name: String,
+) -> Result<(admin_types::PaymentLinkConfig, String), error_stack::Report<errors::ApiErrorResponse>>
+{
+    let (domain_name, business_config) = if let Some(business_config) = business_link_config {
+        let extracted_value: api_models::admin::BusinessPaymentLinkConfig = business_config
+            .parse_value("BusinessPaymentLinkConfig")
+            .change_context(errors::ApiErrorResponse::InvalidDataValue {
+                field_name: "payment_link_config",
+            })
+            .attach_printable("Invalid payment_link_config given in business config")?;
+
+        (
+            extracted_value
+                .domain_name
+                .clone()
+                .map(|d_name| format!("https://{}", d_name))
+                .unwrap_or_else(|| default_domain_name.clone()),
+            Some(extracted_value.config),
+        )
+    } else {
+        (default_domain_name, None)
+    };
+
+    let theme = payment_create_link_config
+        .clone()
```

Comment:
use `as_ref` here and then clone the inner value. You can do the same for a lot of clones below. Fight the compiler